### DISCUSSION TITLE
Run Job in a hook ensuring that pods are not running is not possible

### DISCUSSION THREAD

**Discussion by mgs-garcia** [argoproj/argo-cd#11974] [Q&A]
I need to run a DB schema update before deploying a new App release.
Naturally I though about using a `PreSync` Hook that runs a kubernetes Job with the DB update, but the problem is that my existing pods would keep running while the DB update executes, and that's not desired. Then I thought about setting the previous Hook as `Sync` and also annotate the `Deployment` with `argocd.argoproj.io/hook: Sync`, but with different waves so the Deployment synchronizes before the Job. Plus, I set the Deployment Strategy Type as "Recreate", so the pods are removed before the new ones are span up, and I added an `initContainer` to the Deployment that checks whether the DB update has finished before running the App container. Now the problem is that when a new release of the App happens, it starts by removing existing pods (that's ok), then starts the initContainer that loops until the DB update is ready (blocking the Deployment), but then the DB Update Job never runs because it needs the previous wave to be successful, and a Deployment isn't consider successful until its main containers are running, but they are actually stuck because of the initContainer.

I hope my explanation is not very confusing. In other words, I don't see a way in ArgoCD to run a Job during a deployment and making sure that there are no Pods running at the moment the Job starts executing.

Note that I thought about using `readinessProbe` in the Deployment instead of `initContainer` to check if the DB update has finished, but that's not exactly the same because it ensures that the App doesn't receive new requests but it could still create connections against the DB depending on the App logic.

Finally, I also considered creating another Job as a `PreSync` Hook that scales the replicas down to 0 for the Deployment, but I think it's quite cumbersome and doesn't seem like an elegant solution.

**Comment by jannfis** [comment]
If you'd ask me, I think your consideration of scaling down the deployments in a PreSync hook is the right way to go, and I think quite a few people do it like this. If this PreSync job fails, the sync will not proceed (and that's probably what you want).

I would not suggest to make your Deployment resource a sync hook, because that leads Argo CD to mostly dismiss this resource from the reconciliation loop after it has been run (e.g. it would not have a health check, it would have no self-heal capability, etc).

You can have sync waves on PreSync hooks as well, so you can model a basic flow like

* Run PreSync Job that scales down your deployment to 0 (wave 1)
* Run PreSync job that performs database migration and wait until the (wave 2)
* Run PreSync job that scales up your deployment back to whatever you want it

What do you think is cumbersome and non-elegant about it?

**Reply by mgs-garcia** [reply]
hi @jannfis thank you very much for you answer.
From the 3 steps you mentioned I would eliminate the third, so when the DB upgrade finishes (2nd step), the Argo synchronization process starts and deploys the new version of the app, which will replace the existing `Deployment` with the new release and also spin up the new pods. In other words, I don't think it's necessary to scale the deployment up with a `Job`.

I think it's not elegant because you have to create a `ServiceAccount` and `ClusterRoleBinding` with right permissions to run the `Job` that scales down the service. Plus, I'm using Kustomize with overlays per environment, and I have to annotate each `ServiceAccount` with a different IAM Role depending on the environment, so I think it's cumbersome and makes me doubt if this is the right way to do it. Seems like I have to set up too much stuff to do a simple thing: **stop my service to do a DB upgrade before deploying new releases.** Sounds like a very common use case that, in my humble opinion, should have an easier and straightforward solution instead of all the workaround we're discussing that sounds like a hack or handmade solution.

**Reply by mgs-garcia** [reply]
@jannfis Another issue is that I'll probably be adding HPA (Horizontal Pod Autoscaler) to my solution, so the first step to downscale the pods will be reverted by the HPA... and I wouldn't even think about touching the HPA as part of the deployment procedure; that would definitely be non-elegant :)

**Reply by memelet** [reply]
@mgs-garcia How did you ultimately handle this?
